### PR TITLE
Create tests

### PR DESCRIPTION
**Pull request by ChabVlad**
https://prnt.sc/kzahU7j7vHob

### CODE REVIEW

File: src/test/java/project/bookstore/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    @BeforeEach
+    void setUp(
+            @Autowired DataSource dataSource,
+            @Autowired WebApplicationContext applicationContext
+    ) throws SQLException {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        tearDown(dataSource);
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/controller/BookControllerTest.java]: Why do you need to call tearDown in beforeEach if you already called it in atrerEach method?

File: src/test/java/project/bookstore/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookDto expected = new BookDto();
+        expected.setTitle(requestDto.getTitle());
+        expected.setAuthor(requestDto.getAuthor());
+        expected.setIsbn(requestDto.getIsbn());
+        expected.setPrice(requestDto.getPrice());
+        expected.setCategories(Set.of(new Category()));
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/controller/BookControllerTest.java:168]: I see the same code in the previous method. You can move it to the private method and reuse ;)

File: src/test/java/project/bookstore/controller/CategoryControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    @BeforeEach
+    void setUp(
+            @Autowired DataSource dataSource,
+            @Autowired WebApplicationContext applicationContext
+    ) throws SQLException {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        tearDown(dataSource);
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/controller/CategoryControllerTest.java]: the same

File: src/test/java/project/bookstore/controller/CategoryControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        categoryDto2 = new CategoryDto();
+        categoryDto2.setId(2L);
+        categoryDto2.setName("Adventures");
+
+        categoryDto3 = new CategoryDto();
+        categoryDto3.setId(3L);
+        categoryDto3.setName("Fantasy");
+
+        requestDto = new CategoryRequestDto("Fantasy", null);
+    }
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/controller/CategoryControllerTest.java]: you can move it to the private method and use only one BeforeEach method

File: src/test/java/project/bookstore/controller/CategoryControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            Create category
+            """)
+    @WithMockUser(username = "user", roles = {"ADMIN"})
+    void createCategory_ValidRequestDto_ReturnCategoryDto() throws Exception {
+        String requestJson = objectMapper.writeValueAsString(requestDto);
+        MvcResult result = mockMvc
+                .perform(post(URL)
+                        .content(requestJson)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/controller/CategoryControllerTest.java]: you can use status CREATED for this controller (and other similar methods)

File: src/test/java/project/bookstore/service/BookServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        verify(bookRepository, times(1)).save(book);
+    }
+
+    @Test
+    @DisplayName("""
+            Create a new book with invalid request dto and save it to database
+            """)
+    void save_InvalidRequestDto_ReturnsException() {
+        doThrow(new IllegalArgumentException("Invalid request DTO"))
+                .when(bookMapper).toModel(requestDto);
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/service/BookServiceTest.java]: do you have such logic in your mapper?  I can't find it. So, what are you testing here?

File: src/test/java/project/bookstore/service/BookServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            Update book and save it to database
+            """)
+    void updateBook_ValidRequestDto_ReturnsBookDto() {
+        requestDto.setPrice(BigDecimal.valueOf(12.95));
+
+        doAnswer(invocation -> {
+            Book argBook = invocation.getArgument(1);
+            argBook.setPrice(requestDto.getPrice());
+            return null;
+        })
```
Comment (liliia-ponomarenko) [inline src/test/java/project/bookstore/service/BookServiceTest.java]: why do you need it?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution and answer the question ;)

File: src/test/java/project/bookstore/controller/BookControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                .perform(get(URL).contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        List<BookDto> expected = List.of(bookDto1, bookDto2, bookDto3);
+        List<BookDto> actual = objectMapper.readValue(
+                        result.getResponse().getContentAsString(),
+                        new TypeReference<List<BookDto>>() {});
+
+        assertNotNull(actual);
+        EqualsBuilder.reflectionEquals(expected, actual);
```
Comment (ahoienko) [inline src/test/java/project/bookstore/controller/BookControllerTest.java:98]: Please use static import

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**